library/std/src/sys/net/connection/uefi/tcp4.rs

```diff
@@ -0,0 +1,124 @@
+
+impl Tcp4 {
+    pub(crate) fn new() -> io::Result<Self> {
+        let service_binding = helpers::ServiceProtocol::open(tcp4::SERVICE_BINDING_PROTOCOL_GUID)?;
+        let protocol = helpers::open_protocol(service_binding.child_handle(), tcp4::PROTOCOL_GUID)?;
+
+        Ok(Self { service_binding, protocol, flag: AtomicBool::new(false) })
+    }
+
+    pub(crate) fn configure(
+        &self,
+        active: bool,
+        remote_address: Option<&SocketAddrV4>,
+        station_address: Option<&SocketAddrV4>,
+    ) -> io::Result<()> {
+        let protocol = self.protocol.as_ptr();
+
+        let (remote_address, remote_port) = if let Some(x) = remote_address {
+            (helpers::ipv4_to_r_efi(*x.ip()), x.port())
+        } else {
+            (DEFAULT_ADDR, 0)
+        };
+
+        // FIXME: Remove when passive connections with proper subnet handling are added
+        assert!(station_address.is_none());
+        let use_default_address = efi::Boolean::TRUE;
+        let (station_address, station_port) = (DEFAULT_ADDR, 0);
+        let subnet_mask = helpers::ipv4_to_r_efi(crate::net::Ipv4Addr::new(0, 0, 0, 0));
+
+        let mut config_data = tcp4::ConfigData {
+            type_of_service: TYPE_OF_SERVICE,
+            time_to_live: TIME_TO_LIVE,
+            access_point: tcp4::AccessPoint {
+                use_default_address,
+                remote_address,
+                remote_port,
+                active_flag: active.into(),
+                station_address,
+                station_port,
+                subnet_mask,
+            },
+            control_option: crate::ptr::null_mut(),
+        };
+
+        let r = unsafe { ((*protocol).configure)(protocol, &mut config_data) };
+        if r.is_error() { Err(crate::io::Error::from_raw_os_error(r.as_usize())) } else { Ok(()) }
+    }
+
+    pub(crate) fn connect(&self) -> io::Result<()> {
+        let evt = self.create_evt()?;
+        let completion_token =
+            tcp4::CompletionToken { event: evt.as_ptr(), status: Status::SUCCESS };
+
+        let protocol = self.protocol.as_ptr();
+        let mut conn_token = tcp4::ConnectionToken { completion_token };
+
+        let r = unsafe { ((*protocol).connect)(protocol, &mut conn_token) };
+        if r.is_error() {
+            return Err(io::Error::from_raw_os_error(r.as_usize()));
+        }
+
+        self.wait_for_flag().unwrap();
+
+        if completion_token.status.is_error() {
+            Err(io::Error::from_raw_os_error(completion_token.status.as_usize()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn create_evt(&self) -> io::Result<helpers::OwnedEvent> {
+        self.flag.store(false, Ordering::Relaxed);
+        helpers::OwnedEvent::new(
+            efi::EVT_NOTIFY_SIGNAL,
+            efi::TPL_CALLBACK,
+            Some(toggle_atomic_flag),
+            Some(unsafe { NonNull::new_unchecked(self.flag.as_ptr().cast()) }),
```

Comment:
That makes `wait_for_flag` unsound – there is no guarantee that the structure will not be moved in the time between the calls, which would invalidate this pointer.

Reply:
Should I use Condvar then? Or is there a way to have Atomic on Heap or something?

Reply:
I'd remove the flag, make the notification function a no-op and use `CheckEvent` on the event.

Reply:
The event is of type Signal. The CheckEvent doc states the following

> If Event is of type
EVT_NOTIFY_SIGNAL, then EFI_INVALID_PARAMETER is returned.

The reason the event is of type Signal is because the tcp4 protocol connect doc states the following

> The Event to signal after request is finished and Status field is updated by the EFI TCPv4 Protocol driver. The
type of Event must be EVT_NOTIFY_SIGNAL, and its Task Priority Level (TPL) must be lower than or equal to
TPL_CALLBACK.

Reply:
Ah, I missed that. Then the easiest way is probably just to ensure that the flag outlives the event. Why not just inline `wait_for_flag` and make `flag` a local variable? Alternatively, make `create_evt` `unsafe`.

Reply:
> Ah, I missed that. Then the easiest way is probably just to ensure that the flag outlives the event. Why not just inline `wait_for_flag` and make `flag` a local variable? Alternatively, make `create_evt` `unsafe`.

I would prefer not to inline `wait_for_flag` since that will be repeated in almost all other implementations soon enough (read, write, close, etc). I have made `create_evt` unsafe. Is that ok?

Reply:
It's not the cleanest solution ever™, but I guess it's fine. In the future it might be nice to create a single global event for all "blocking" I/O calls in `std`.